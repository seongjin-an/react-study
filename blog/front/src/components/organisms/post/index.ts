export {PostViewer} from './PostViewer'
export {PostContent} from './PostContent'