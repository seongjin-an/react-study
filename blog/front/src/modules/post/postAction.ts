import {createAction} from "@reduxjs/toolkit";
import {READ_POST_REQUEST, UNLOAD_POST} from "./postType";

export const readPostAction = createAction<string>(READ_POST_REQUEST)
export const unloadPostAction = createAction(UNLOAD_POST)