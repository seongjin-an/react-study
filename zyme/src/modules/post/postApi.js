import axios from "axios";

export const getList = () => {
    return axios.get('https://jsonplaceholder.typicode.com/posts')
}