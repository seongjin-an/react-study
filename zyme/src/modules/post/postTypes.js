export const READ_POST_REQUEST = 'post/READ_POST_REQUEST'
export const READ_POST_SUCCESS = 'post/READ_POST_SUCCESS'
export const READ_POST_FAILURE = 'post/READ_POST_FAILURE'