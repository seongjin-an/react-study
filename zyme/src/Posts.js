import React, {useEffect} from "react";
import {useDispatch, useSelector} from "react-redux";
import {readPostRequest} from "./modules/post/postAction";

const Posts = () => {
    const dispatch = useDispatch()
    const list = useSelector((state) => state.post.list)
    useEffect(() => {
        dispatch(readPostRequest())
    },[])
    return(
        <div className="postArea">
            {list && list.length>0 &&
            list.map((item, index) => <div key={index}>
                <div>{item.userId}</div>
                <div>{item.id}</div>
                <div>{item.title}</div>
                <div>{item.body}</div>
            </div>)
            }
        </div>
    )
}
export default Posts