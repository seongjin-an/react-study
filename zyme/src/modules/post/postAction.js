import {createAction} from "@reduxjs/toolkit";
import {READ_POST_FAILURE, READ_POST_REQUEST, READ_POST_SUCCESS} from "./postTypes";

export const readPostRequest = createAction(READ_POST_REQUEST)
export const readPostSuccess = createAction(READ_POST_SUCCESS)
export const readPostFailure = createAction(READ_POST_FAILURE)