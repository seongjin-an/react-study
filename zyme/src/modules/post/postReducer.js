import {createReducer} from "@reduxjs/toolkit";
import {READ_POST_FAILURE, READ_POST_SUCCESS} from "./postTypes";

const initialState = {
    list: [],
    error: false
}

const post = createReducer(initialState, {
    [READ_POST_SUCCESS]: (state, action) => {
        return{
            ...state,
            list: action.payload
        }
    },
    [READ_POST_FAILURE]: (state, action) => {
        return{
            ...state,
            error: true
        }
    }
})

export default post