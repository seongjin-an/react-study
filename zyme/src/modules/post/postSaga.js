import {all, call, fork, put, takeLatest} from "@redux-saga/core/effects";
import axios from "axios";
import {readPostFailure, readPostSuccess} from "./postAction";
import {READ_POST_REQUEST} from "./postTypes";

const fetchReadPostRequest = function*(){
    try{
        const getList = () => axios.get('https://jsonplaceholder.typicode.com/posts')
        const response = yield call(getList)
        if(response.data){
            yield put(readPostSuccess(response.data))
        }else{
            yield put(readPostFailure())
        }
    }catch(error){
        yield put(readPostFailure())
    }
}

const watchReadPostRequest = function*(){
    yield takeLatest(READ_POST_REQUEST, fetchReadPostRequest)
}

const postSaga = function*(){
    yield all([
        fork(watchReadPostRequest)
    ])
}
export default postSaga