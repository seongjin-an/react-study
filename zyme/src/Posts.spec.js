import React from "react";
import {configure, mount} from "enzyme";
import {Provider} from "react-redux";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import Posts from "./Posts";
import toJson from "enzyme-to-json";
import "@testing-library/jest-dom/extend-expect"
import configureMockStore from 'redux-mock-store'
import * as actions from "./modules/post/postAction";
import * as redux from 'react-redux'
import {READ_POST_REQUEST} from "./modules/post/postTypes";
import axios from "axios";
import * as api from "./modules/post/postApi";

let store
const setup2 = (WrappedComponent, props={}) => {
    const wrapper = mount(
        <Provider store={store}>
            <WrappedComponent {...props}/>
        </Provider>
    )
    return {wrapper}
}

const mockStore = configureMockStore()
store = mockStore()
jest.mock('axios')
describe('enzyme saga test', () => {
    let spyOnUseSelector;
    let spyOnUseDispatch;
    let mockDispatch;
    beforeEach(() => {
        store = mockStore({
            post: {
                list: [
                    {"userId": 11, "id": 11, "title": 'hello', "body": 'world'}
                ]
            }
        })
        configure({ adapter: new Adapter() });
    })
    afterEach(() => {
        jest.restoreAllMocks();
    });
    it('should match snapshot', () => {
        // const wrapper = render(<Provider store={store}><Posts/></Provider>)
        // expect(await screen.findByText('sunt aut facere repellat provident occaecati excepturi optio reprehenderit')).toBeInTheDocument()

        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        // // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();

        spyOnUseSelector.mockImplementation((selectorFn) => selectorFn({
            post:{ list: [{userId: 2, id: 2, title: '2', body: '22'}]}
        }));
        spyOnUseDispatch.mockReturnValue(mockDispatch);
        const { wrapper } = setup2(Posts)
        expect(toJson(wrapper)).toMatchSnapshot()
   })
    it('should match actions', () => {
        const expectedActions = {
            readPostRequest:{
                'payload': undefined,
                'type': 'post/READ_POST_REQUEST',
            },
            readPostSuccess:{
                'payload': [{userId:1, id:1, title: 'hello', body: 'world'}],
                'type': 'post/READ_POST_SUCCESS'
            },
            readPostFailure:{
                'payload': undefined,
                'type': 'post/READ_POST_FAILURE'
            }
        };
        expect(actions.readPostRequest()).toEqual(expectedActions.readPostRequest)
        expect(actions.readPostSuccess([{userId:1,id:1,title:'hello',body:'world'}])).toEqual(expectedActions.readPostSuccess)
        expect(actions.readPostFailure()).toEqual(expectedActions.readPostFailure)
    });

    it('should spy on useSelector and useDispatch', () => {
        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        spyOnUseSelector.mockReturnValue({
            post: {
                list: [
                    {"userId": 11, "id": 11, "title": 'hello', "body": 'world'}
                ]
            }
        });
        //
        // // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();
        spyOnUseDispatch.mockReturnValue(mockDispatch);
        const { wrapper } = setup2(Posts)

        expect(spyOnUseSelector).toBeCalledTimes(1);
        expect(spyOnUseDispatch).toBeCalledTimes(1);
        actions.readPostRequest(mockDispatch)
        expect(mockDispatch.mock.calls.length).toBe(1)
        expect(mockDispatch.mock.calls[0][0]).toEqual({
            type: READ_POST_REQUEST,
            value: undefined
        })
        console.log('...................mockDispatch.mock.calls:', mockDispatch.mock.calls)
        console.log('...................store.getActions():', store.getActions())
    })

    it("should return", async () => {
        const spyOnGetList = jest.spyOn(api, "getList");
        spyOnGetList.mockReturnValue({
            data: [
                {
                    userId: 1,
                    id: 1,
                    title: 'My First Album'
                },
                {
                    userId: 1,
                    id: 2,
                    title: 'Album: The Sequel'
                }
            ]
        })
        // axios.get.mockResolvedValue({
        //     data: [
        //         {
        //             userId: 1,
        //             id: 1,
        //             title: 'My First Album'
        //         },
        //         {
        //             userId: 1,
        //             id: 2,
        //             title: 'Album: The Sequel'
        //         }
        //     ]
        // });
        const list = await api.getList();
        console.log('list....:', list);
        /*
        const user = {
      login: 'user 1',
      password: 'password'
    };
    const dispatch = jest.fn();
    await actions.signInUser(user)(dispatch);  // <= call the function on a user, then call the resulting function on a dispatch
    expect(dispatch).toHaveBeenCalledTimes(2);  // Success!
    expect(dispatch).toHaveBeenNthCalledWith(1, { type: START_SIGNIN_REQUEST });  // Success!
    expect(dispatch).toHaveBeenNthCalledWith(2, { type: SIGNIN_USER_SUCCEEDED, user: 'the username' });  // Success!
         */
    })

    it("should spy", async () => {
        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        // // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();

        spyOnUseSelector.mockImplementation((selectorFn) => selectorFn({
            post:{ list: [{userId: 2, id: 2, title: '2', body: '22'}]}
        }));
        spyOnUseDispatch.mockReturnValue(mockDispatch);
        const { wrapper } = setup2(Posts)
        expect(mockDispatch).toHaveBeenCalledWith({
            'payload': undefined,
            'type': 'post/READ_POST_REQUEST',
        });

        console.log('wrapper::::', toJson(wrapper));
    })
})