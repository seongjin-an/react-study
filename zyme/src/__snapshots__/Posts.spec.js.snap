// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme saga test should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Posts>
    <div>
      <div
        key="0"
      >
        <div>
          2
        </div>
        <div>
          2
        </div>
        <div>
          2
        </div>
        <div>
          22
        </div>
      </div>
    </div>
  </Posts>
</Provider>
`;
